eth/block.go

```diff
@@ -83,6 +87,71 @@ func (b *Block) UnmarshalJSON(data []byte) error {
 func (b Block) MarshalJSON() ([]byte, error) {
 	switch b.flavor {
 	case "geth":
+		// Note that post EIP-4895 we have to behave slightly differently if WithdrawalsRoot is set.
+		// if it's set, then `Withdrawals` should be non-nil
+		// if it's not set, then Withdrawals should be omitted
+		if b.WithdrawalsRoot != nil {
+			type withWithdrawals struct {
```

Comment:
It _might_ be a good idea to declare these structs outside of the function so that methods can be defined on the type in the future. It also would make it easier to test, though that change would break with the existing pattern that exists here.